File: Chapter12/scala/com/packtpub/spark/module_four/chapter_12/streaming/TransformingStreams.scala
package com.packtpub.spark.module_four.chapter_12.streaming

import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.spark.sql.SparkSession
import org.apache.spark.streaming.kafka010.{ConsumerStrategies, KafkaUtils, LocationStrategies}
import org.apache.spark.streaming.{Seconds, StreamingContext}

object TransformingStreams {

  def main(args: Array[String]): Unit = {

    // Build a SparkSession in Local Mode
    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("My Spark App")
      .getOrCreate()

    val sc = spark.sparkContext

    // Create a StreamingContext
    val ssc = new StreamingContext(sc, Seconds(10))

    // Create a list of servers
    val servers = Seq("localhost:9092").mkString(",")

    val kafka_params = Map[String, Object](
      "bootstrap.servers" -> servers,
      "key.deserializer" -> classOf[StringDeserializer],
      "value.deserializer" -> classOf[StringDeserializer],
      "group.id" -> "streaming_test",
      "auto.offset.reset" -> "latest",
      "enable.auto.commit" -> (false: java.lang.Boolean))

    val topics = Array("website_traffic")
    val strategy = ConsumerStrategies.Subscribe[String, String](topics, kafka_params)

    val rawData = KafkaUtils.createDirectStream[String, String](
      ssc,
      LocationStrategies.PreferConsistent,
      strategy
    )

    // Create a dictionary/lookup of actions (int) to their description
    val actions = Seq((1, "opened website"), (2, "clicked"), (3, "scrolled"))
    val webActions =  spark.sparkContext.parallelize(actions)

    // Turn the raw stream data into
    val parsedWebTraffic = rawData.map(x => {
      val parts = x.value().split(",")
      val userID = parts{0}
      val actionID = parts{1}.toInt
      (actionID, userID)
    })


    parsedWebTraffic.print()

    // Use transform method to join data with lookup
    val enhanced = parsedWebTraffic.transform(x => {
      x.join(webActions)
    })
    enhanced.print()

    // Allow the streaming context to start, and stick around until terminated
    ssc.start()
    ssc.awaitTermination()
  }
}

File: Chapter12/python/spark_streaming/TransformingStreams.py
from pyspark import SparkContext
from pyspark.streaming import StreamingContext
from pyspark.streaming.kafka import KafkaUtils

sc = SparkContext("local[2]", "My Spark App")
ssc = StreamingContext(sc, 10)


def read(data): print(data)


kafka_params = {"bootstrap.servers": "localhost:9092",
                "key.deserializer": "org.apache.kafka.common.serialization.StringDeserializer",
                "value.deserializer": "org.apache.kafka.common.serialization.StringDeserializer",
                "group.id": "streaming_test",
                "auto.offset.reset": "largest",
                "enable.auto.commit": "False"}

topics = ['website_traffic']

kafkaStream = KafkaUtils.createDirectStream(ssc=ssc, topics=topics, kafkaParams=kafka_params)

actions = [(1, "opened website"), (2, "clicked"), (3, "scrolled")]
webActions = sc.parallelize(actions)

kafkaStream.pprint()


# Function to Parse Raw Web Traffic
def parse(record):
    userid_and_actionid = record[1]
    parts = userid_and_actionid.split(",")
    user_id = parts[0]
    action_id = parts[1]
    return int(action_id), user_id


def join(record):
    return record.join(webActions)


# parse the data into a clean format
parsed_web_traffic = kafkaStream.map(parse)
parsed_web_traffic.pprint()

# join the data
enhanced = parsed_web_traffic.transform(join)
enhanced.pprint()

ssc.start()
ssc.awaitTermination()
